docs/cli/ilab-model-serve.md

```diff
@@ -0,0 +1,101 @@
+
+## Design
+
+Since the subject of the `ilab model serve` command is a model, regardless of the format of the model, every command takes in the `--model` flag or uses its default value in the config.
+
+`ilab model serve` has a subcommand for each backend it can serve models with. As of this draft, the two backends `ilab` would serve with are `llama-cpp` and `vllm`.
+
+This would lead to the commands:
+
+- `ilab model serve llama-cpp`
+- `ilab model serve vllm`
+
+The flags for `ilab model serve` are either backend agnostic or specific to a subcommand and are prefixed to indicate that. The intention behind this choice is to ensure there is no mix up of flags between backends.
+
+```console
+ilab model serve
+|
+|_______ --model
+|
+|_______ (other backend agnostic flags)
+|
+|_______ llama-cpp
+|        |
+|        |_______ --llama-cpp-num-gpus
+|        |
+|        |_______ --llama-cpp-max-ctx-size
+|        |
+|        |_______ (other llama-cpp specific args)
+|
+|_______ vllm
+|        |
+|        |_______ --vllm-chat-template
+|        |
+|        |_______ (other popular vllm-specific specific args)
+|        |
+|        |_______ --vllm-args
+```
+
+## Command Examples
+
+### Bare-bones but model specific command
+
+```shell
+ilab model serve --model <PATH> 
+```
+
+- Serves the model at `<PATH>`.
+- If the `<PATH>` is the path for a `.gguf` file then `llama-cpp` is automatically used as the model serving backend.
+- If the path is a directory, `vllm` is automatically used as the backend.
+- The path provided to `vllm` can either be to a Hugging Face repo or a directory with `.safetensors`.
+
+### Bare-bones command
+
+```shell
+ilab model serve
+```
+
+- This command has the same behavior as the one above but the `--model-path` is whatever the default model path is in the config. This is the existing behavior of `ilab serve` today.
+
+### Llama-cpp backend specific commands
+
+```shell
+ilab model serve llama-cpp --model <PATH> --llama-cpp-gpu-layers 4 --llama-cpp-max-ctx-size 1024
+```
+
+- This command serves a model with `llama-cpp`.
+- If the path provided is not to a `.gguf` file, this command would error out.
+- The existing flags to `ilab serve` (besides `--model-path` & `--log-file`) are now specific to the llama-cpp backend and are prefixed with `--llama-cpp-` to indicate that.
+
+### vllm backend specific commands
+
+```shell
+ilab model serve vllm --model <PATH> --vllm-chat-template <PATH> --vllm-args "<MORE VLLM SPECIFIC ARGS>"
+```
+
+- This command serves a model with `vllm`.
+- If the path provided is not a directory, this command would error out.
+- There are [dozens](https://docs.vllm.ai/en/latest/serving/openai_compatible_server.html#command-line-arguments-for-the-server) of flags for vllm. Whichever arguments the community deems the most important to include, will be added as flags to `ilab model serve vllm` with `--vllm-` as a prefix.
+- Any remaining arguments can be specified in the value of the flag `--vllm-args`.
```

Comment:
The only issue with this approach is how can we easily translate a given arg to an internal value? With a flag that is very easy to do. The downside with the flag is that we keep playing catching with every flag vllm produces...

Reply:
I don't think I'm being very clear here, the intention of `--vllm-args` is to not play catch up with every flag vllm has, but instead is intended to be a catch-all flag for all of it's arguments, which is something I think you'd support.

For example the community deams the [vllm arguments](https://docs.vllm.ai/en/latest/serving/openai_compatible_server.html#command-line-arguments-for-the-server) `--chat-template` and `--download-dir` to be ones that we use often and want to have flags for the subcommand. 

Those would get prefixed with `--vllm-` and become `--vllm-chat-template` and `--vllm-download-dir`. 

If a user wanted to specify more "obscure" arguments to vllm like the `--kv-cache-dtype` and `--block-size` when serving vllm, an example of a command would be:

`ilab model serve vllm --vllm-download-dir <PATH> --vllm-args "--kv-cache-dtype auto --block-size 16"`.

The value of `--vllm-args` is just getting passed straight into the vllm upon startup. The main function of the openai compatible server take and processes all arguments on start up.

Reply:
Another example of why this is necessary is that a user also needs to sometimes use hidden behavioral flags that do not have a long lifespan to work around a problem in the backend. Yet another case would be cuda flags that vllm needs to pass through

Reply:
I like the catchall approach. It's one we used to handle arguments to plugins we don't control in the ChRIS project. It has worked well.

Reply:
Ok now it's clearer, I'm ok with the `--vllm-args` value being passed as is straight upon vllm startup.